core/src/main/scala/kafka/log/FileMessageSet.scala

```diff
@@ -149,6 +155,67 @@ class FileMessageSet private[kafka](@volatile var file: File,
   }
 
   /**
+   * Search forward for the message whose timestamp is greater than or equals to the target timestamp.
+   *
+   * The search will stop immediately when it sees a message in format version before 0.10.0. This is to avoid
+   * scanning the entire log when all the messages are still in old format.
+   *
+   * @param targetTimestamp The timestamp to search for.
+   * @param startingPosition The starting position to search.
+   * @return None, if no message exists at or after the starting position.
+   *         Some(TimestampOffset) with (largest_timestamp_checked, the_next_offset_to_read) otherwise.
+   */
+  def searchForTimestamp(targetTimestamp: Long, startingPosition: Int): Option[TimestampOffset] = {
+    var maxTimestampChecked = Message.NoTimestamp
+    var lastOffsetChecked = -1L
+    val messagesToSearch = read(startingPosition, sizeInBytes)
+    for (messageAndOffset <- messagesToSearch) {
+      val message = messageAndOffset.message
+      lastOffsetChecked = messageAndOffset.offset
+      // Stop searching once we see message format before 0.10.0.
+      // This equivalent as treating message without timestamp has the largest timestamp.
+      // We do this to avoid scanning the entire log if no message has a timestamp.
+      if (message.magic == Message.MagicValue_V0)
+        return Some(TimestampOffset(maxTimestampChecked, messageAndOffset.offset))
```

Comment:
Thanks for the explanation. 

10. Yes, I agree that it's better to use last modified time for segments with no timestamp. So, if all messages have NoTimestamp, the semantic is that the search will return the offset of the first message in the segment whose last modified time is >= the target timestamp.

11. If a segment has messages with and w/o timestamp, to find the message matching the precise timestamp may require us scan most the segment, but is probably ok since it's rare.

It would be useful to document the semantics in the comment above Log.fetchOffsetsByTimestamp() on those corner cases.